src/sentry/search/snuba/backend.py

```diff
@@ -341,6 +285,54 @@ def _query(self, projects, retention_window_start, group_queryset, tags, environ
         max_time = options.get('snuba.search.max-total-chunk-time-seconds')
         time_start = time.time()
 
+        if count_hits and candidate_ids is None:
+            # If we have no candidates, get a random sample of groups matching
+            # the snuba side of the query, and see how many of those pass the
+            # post-filter in postgres. This should give us an estimate of the
+            # total number of snuba matches that will be overall matches, which
+            # we can use to get an estimate for X-Hits. Note no cursor, so we
+            # are always estimating the total hits.
+
+            # The number of samples required to achieve a certain error bound
+            # with a certain confidence interval can be calculated from a
+            # rearrangement of the normal approximation (Wald) confidence
+            # interval formula:
+            #
+            # https://en.wikipedia.org/wiki/Binomial_proportion_confidence_interval
+            #
+            # Effectively if we want the estimate to be within +/- 10% of the
+            # real value with 95% confidence, we would need (1.96^2 * p*(1-p))
+            # / 0.1^2 samples. With a starting assumption of p=0.5 (this
+            # requires the most samples) we would need 96 samples to achieve
+            # +/-10% @ 95% confidence.
+
+            sample_size = options.get('snuba.search.hits-sample-size')
+            snuba_groups, snuba_total = snuba_search(
+                start=start,
+                end=end,
+                project_ids=[p.id for p in projects],
+                environment_ids=environments and [environment.id for environment in environments],
+                tags=tags,
+                sort_field=sort_field,
+                limit=sample_size,
+                offset=0,
+                get_sample=True,
+                **parameters
+            )
+            snuba_count = len(snuba_groups)
+            if snuba_count == 0:
+                return EMPTY_RESULT
+            else:
+                filtered_count = group_queryset.filter(
+                    id__in=[gid for gid, _ in snuba_groups]
+                ).count()
+
+                hit_ratio = filtered_count / float(snuba_count)
+                hits = int(hit_ratio * snuba_total)
+                logger.info("estimating hits from snuba_samples: {}, snuba_total: {}, django_hits: {}, hit_ratio: {:.3f}, estimated_hits: {}".format(
```

Comment:
@JTCunning will probably not like this log line.

Reply:
oh yeah this was just for debugging, will remove